File: Creative Organizer.app/Contents/Resources/Organizador.m
#import <Cocoa/Cocoa.h>

typedef NS_ENUM(NSInteger, OrganizerAction) {
    OrganizerActionOrganize,
    OrganizerActionPreview,
    OrganizerActionUndo,
    OrganizerActionCancel,
};

static void ShowMessage(NSString *title, NSString *message) {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.messageText = title;
    alert.informativeText = message;
    [alert addButtonWithTitle:@"Cerrar"];
    [alert runModal];
}

static NSURL *ChooseLaunchFolder(void) {
    NSOpenPanel *panel = [NSOpenPanel openPanel];
    panel.title = @"Organizador de Lanzamientos";
    panel.message = @"Elegi la carpeta del lanzamiento";
    panel.prompt = @"Elegir carpeta";
    panel.canChooseFiles = NO;
    panel.canChooseDirectories = YES;
    panel.canCreateDirectories = NO;
    panel.allowsMultipleSelection = NO;
    panel.directoryURL = [[NSFileManager defaultManager] URLsForDirectory:NSDesktopDirectory inDomains:NSUserDomainMask].firstObject;
    [panel center];
    return [panel runModal] == NSModalResponseOK ? panel.URL : nil;
}

static OrganizerAction ChooseAction(void) {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.messageText = @"Organizador de Lanzamientos";
    alert.informativeText = @"Que queres hacer?";
    [alert addButtonWithTitle:@"Ordenar"];
    [alert addButtonWithTitle:@"Preview"];
    [alert addButtonWithTitle:@"Deshacer el ordenamiento anterior"];
    [alert addButtonWithTitle:@"Salir"];

    switch ([alert runModal]) {
        case NSAlertFirstButtonReturn: return OrganizerActionOrganize;
        case NSAlertSecondButtonReturn: return OrganizerActionPreview;
        case NSAlertThirdButtonReturn: return OrganizerActionUndo;
        default: return OrganizerActionCancel;
    }
}

static NSScrollView *PreviewView(NSString *output) {
    NSRect frame = NSMakeRect(0, 0, 760, 420);
    NSScrollView *scrollView = [[NSScrollView alloc] initWithFrame:frame];
    scrollView.hasVerticalScroller = YES;
    scrollView.hasHorizontalScroller = YES;
    scrollView.autohidesScrollers = YES;
    scrollView.borderType = NSBezelBorder;

    NSTextView *textView = [[NSTextView alloc] initWithFrame:frame];
    textView.editable = NO;
    textView.selectable = YES;
    textView.font = [NSFont monospacedSystemFontOfSize:11 weight:NSFontWeightRegular];
    textView.string = output.length ? output : @"No se detectaron movimientos.";
    textView.minSize = NSMakeSize(0, frame.size.height);
    textView.maxSize = NSMakeSize(CGFLOAT_MAX, CGFLOAT_MAX);
    textView.verticallyResizable = YES;
    textView.horizontallyResizable = YES;
    textView.textContainer.containerSize = NSMakeSize(CGFLOAT_MAX, CGFLOAT_MAX);
    textView.textContainer.widthTracksTextView = NO;
    scrollView.documentView = textView;
    return scrollView;
}

static void ShowPreview(NSString *output) {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.messageText = @"Preview del ordenamiento";
    alert.informativeText = @"Revisa las rutas propuestas. No se movio ningun archivo.";
    alert.accessoryView = PreviewView(output);
    [alert addButtonWithTitle:@"Cerrar"];
    [alert runModal];
}

static BOOL ConfirmPlan(NSString *output) {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.messageText = @"Confirmar ordenamiento";
    alert.informativeText = @"Revisa el plan completo antes de mover archivos.";
    alert.accessoryView = PreviewView(output);
    [alert addButtonWithTitle:@"Ordenar"];
    [alert addButtonWithTitle:@"Cancelar"];
    return [alert runModal] == NSAlertFirstButtonReturn;
}

static NSURL *LastLaunchRecordURL(void) {
    NSError *error = nil;
    NSURL *supportURL = [[NSFileManager defaultManager] URLForDirectory:NSApplicationSupportDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:&error];
    if (!supportURL) return nil;

    NSURL *folderURL = [supportURL URLByAppendingPathComponent:@"Organizador de Lanzamientos" isDirectory:YES];
    if (![[NSFileManager defaultManager] createDirectoryAtURL:folderURL withIntermediateDirectories:YES attributes:nil error:&error]) {
        return nil;
    }
    return [folderURL URLByAppendingPathComponent:@"last-launch.txt"];
}

static BOOL SaveLastLaunchFolder(NSURL *folder) {
    NSURL *recordURL = LastLaunchRecordURL();
    if (!recordURL) return NO;
    return [folder.path writeToURL:recordURL atomically:YES encoding:NSUTF8StringEncoding error:nil];
}

static NSURL *LoadLastLaunchFolder(void) {
    NSURL *recordURL = LastLaunchRecordURL();
    if (!recordURL) return nil;
    NSString *path = [NSString stringWithContentsOfURL:recordURL encoding:NSUTF8StringEncoding error:nil];
    if (!path.length) return nil;

    BOOL isDirectory = NO;
    if (![[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDirectory] || !isDirectory) {
        return nil;
    }
    return [NSURL fileURLWithPath:path];
}

static void ClearLastLaunchFolder(void) {
    NSURL *recordURL = LastLaunchRecordURL();
    if (recordURL) {
        [[NSFileManager defaultManager] removeItemAtURL:recordURL error:nil];
    }
}

static NSDictionary *RunOrganizer(NSURL *folder, NSString *flag) {
    NSURL *resourcesURL = [NSBundle mainBundle].resourceURL;
    NSString *script = [resourcesURL URLByAppendingPathComponent:@"ordenar_lanzamiento.py"].path;
    NSURL *pythonURL = [resourcesURL URLByAppendingPathComponent:@"Python3.framework/Versions/Current/bin/python3"];
    if (!script || ![[NSFileManager defaultManager] isExecutableFileAtPath:pythonURL.path]) {
        return @{@"success": @NO, @"output": @"No encontre el archivo interno del organizador."};
    }

    NSTask *task = [[NSTask alloc] init];
    NSPipe *pipe = [NSPipe pipe];
    // Python travels inside the app, so teammates do not need Xcode or Python installed.
    task.executableURL = pythonURL;
    task.arguments = @[script, @"--root", folder.path, flag];
    task.standardOutput = pipe;
    task.standardError = pipe;

    NSError *error = nil;
    if (![task launchAndReturnError:&error]) {
        return @{@"success": @NO, @"output": error.localizedDescription ?: @"No pude ejecutar el organizador."};
    }
    NSData *data = [pipe.fileHandleForReading readDataToEndOfFile];
    [task waitUntilExit];
    NSString *output = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    output = [output stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    return @{@"success": @(task.terminationStatus == 0), @"output": output ?: @""};
}

static NSString *MarketsFromOutput(NSString *output) {
    for (NSString *line in [output componentsSeparatedByString:@"\n"]) {
        if ([line hasPrefix:@"Mercados detectados:"]) {
            return line;
        }
    }
    return @"Mercados detectados: ninguno";
}

int main(void) {
    @autoreleasepool {
        [NSApplication sharedApplication];
        [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
        [NSApp activateIgnoringOtherApps:YES];

        OrganizerAction action = ChooseAction();
        if (action == OrganizerActionCancel) return 0;

        if (action == OrganizerActionUndo) {
            NSURL *folder = LoadLastLaunchFolder();
            if (!folder) {
                ClearLastLaunchFolder();
                ShowMessage(@"No hay un ordenamiento para deshacer", @"Organiza una carpeta primero. La app solo puede deshacer el ultimo ordenamiento hecho desde esta app.");
                return 0;
            }
            NSDictionary *result = RunOrganizer(folder, @"--undo");
            if ([result[@"success"] boolValue] || [result[@"output"] containsString:@"No encontre un undo"]) {
                ClearLastLaunchFolder();
            }
            ShowMessage([result[@"success"] boolValue] ? @"Undo aplicado" : @"No se pudo hacer Undo", [result[@"success"] boolValue] ? @"La carpeta volvio al estado anterior." : result[@"output"]);
            return 0;
        }

        NSURL *folder = ChooseLaunchFolder();
        if (!folder) return 0;

        NSDictionary *preview = RunOrganizer(folder, @"--preview");
        if (![preview[@"success"] boolValue]) {
            ShowMessage(@"No se pudo generar el preview", preview[@"output"]);
            return 0;
        }
        if (action == OrganizerActionPreview) {
            ShowPreview(preview[@"output"]);
            return 0;
        }
        if (!ConfirmPlan(preview[@"output"])) return 0;

        NSDictionary *result = RunOrganizer(folder, @"--apply");
        if ([result[@"success"] boolValue]) {
            SaveLastLaunchFolder(folder);
        }
        NSString *message = [result[@"success"] boolValue] ? [NSString stringWithFormat:@"Listo. %@", MarketsFromOutput(result[@"output"])] : result[@"output"];
        ShowMessage([result[@"success"] boolValue] ? @"Organizacion terminada" : @"No se pudo ordenar", message);
    }
    return 0;
}

File: Creative Organizer.app/Contents/Resources/Organizador.swift
import Cocoa

enum OrganizerAction {
    case organize
    case preview
    case undo
}

func showMessage(title: String, message: String) {
    let alert = NSAlert()
    alert.messageText = title
    alert.informativeText = message
    alert.addButton(withTitle: "Cerrar")
    alert.runModal()
}

func chooseLaunchFolder() -> URL? {
    let panel = NSOpenPanel()
    panel.title = "Organizador de Lanzamientos"
    panel.message = "Elegi la carpeta del lanzamiento"
    panel.prompt = "Elegir carpeta"
    panel.canChooseFiles = false
    panel.canChooseDirectories = true
    panel.canCreateDirectories = false
    panel.allowsMultipleSelection = false
    panel.directoryURL = FileManager.default.urls(for: .desktopDirectory, in: .userDomainMask).first
    panel.center()

    return panel.runModal() == .OK ? panel.url : nil
}

func chooseAction(folderName: String) -> OrganizerAction? {
    let alert = NSAlert()
    alert.messageText = folderName
    alert.informativeText = "Que queres hacer con esta carpeta?"
    alert.addButton(withTitle: "Ordenar")
    alert.addButton(withTitle: "Preview")
    alert.addButton(withTitle: "Undo")
    alert.addButton(withTitle: "Cancelar")

    switch alert.runModal() {
    case .alertFirstButtonReturn: return .organize
    case .alertSecondButtonReturn: return .preview
    case .alertThirdButtonReturn: return .undo
    default: return nil
    }
}

func confirm(title: String, message: String, button: String) -> Bool {
    let alert = NSAlert()
    alert.messageText = title
    alert.informativeText = message
    alert.addButton(withTitle: button)
    alert.addButton(withTitle: "Cancelar")
    return alert.runModal() == .alertFirstButtonReturn
}

func runOrganizer(root: URL, flag: String) -> (success: Bool, output: String) {
    guard let scriptURL = Bundle.main.resourceURL?.appendingPathComponent("ordenar_lanzamiento.py") else {
        return (false, "No encontre el archivo interno del organizador.")
    }

    let process = Process()
    let outputPipe = Pipe()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/python3")
    process.arguments = [scriptURL.path, "--root", root.path, flag]
    process.standardOutput = outputPipe
    process.standardError = outputPipe

    do {
        try process.run()
        process.waitUntilExit()
        let data = outputPipe.fileHandleForReading.readDataToEndOfFile()
        let output = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return (process.terminationStatus == 0, output)
    } catch {
        return (false, "No pude ejecutar el organizador: \(error.localizedDescription)")
    }
}

func summary(from output: String) -> String {
    let lines = output.split(separator: "\n").filter {
        $0.hasPrefix("Carpeta:") ||
        $0.hasPrefix("Archivos detectados:") ||
        $0.hasPrefix("Archivos listos para mover:") ||
        $0.hasPrefix("Archivos sin resolver:") ||
        $0.hasPrefix("Mercados detectados:")
    }
    return lines.isEmpty ? output : lines.joined(separator: "\n")
}

let app = NSApplication.shared
app.setActivationPolicy(.regular)
app.activate(ignoringOtherApps: true)

guard let folder = chooseLaunchFolder() else {
    app.terminate(nil)
    exit(0)
}

guard let action = chooseAction(folderName: folder.lastPathComponent) else {
    app.terminate(nil)
    exit(0)
}

switch action {
case .preview:
    let result = runOrganizer(root: folder, flag: "--preview")
    showMessage(title: result.success ? "Preview" : "No se pudo hacer el preview", message: summary(from: result.output))

case .undo:
    if confirm(title: "Deshacer organizacion", message: "Esto revierte la ultima organizacion hecha en esta carpeta.", button: "Deshacer") {
        let result = runOrganizer(root: folder, flag: "--undo")
        showMessage(title: result.success ? "Undo aplicado" : "No se pudo hacer Undo", message: result.success ? "La carpeta volvio al estado anterior." : result.output)
    }

case .organize:
    let preview = runOrganizer(root: folder, flag: "--preview")
    guard preview.success else {
        showMessage(title: "No se pudo leer la carpeta", message: preview.output)
        break
    }
    let message = summary(from: preview.output) + "\n\nVoy a mover y renombrar archivos dentro de esta carpeta, sin duplicarlos."
    if confirm(title: "Confirmar organizacion", message: message, button: "Ordenar") {
        let result = runOrganizer(root: folder, flag: "--apply")
        showMessage(title: result.success ? "Organizacion terminada" : "No se pudo ordenar", message: result.success ? "Listo. Los archivos fueron reordenados sin duplicarse." : result.output)
    }
}

app.terminate(nil)
